Sources/SentryCrash/Recording/Tools/SentryCrashCxaThrowSwapper.c

```diff
@@ -0,0 +1,423 @@
+#include <mach-o/nlist.h>
+#include <mach/mach.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+
+#include "SentryAsyncSafeLog.h"
+#include "SentryCrashMach-O.h"
+#include "SentryCrashPlatformSpecificDefines.h"
+
+typedef struct {
+    uintptr_t image_dli_fbase_address;
+    uintptr_t cxa_throw_original_function;
+} SentryCrashImageToOriginalCxaThrowPair;
+
+static cxa_throw_type g_cxa_throw_handler = NULL;
+static const char *const g_cxa_throw_name = "__cxa_throw";
+
+static SentryCrashImageToOriginalCxaThrowPair *g_cxa_originals = NULL;
+static size_t g_cxa_originals_capacity = 0;
+static size_t g_cxa_originals_count = 0;
+
+static uintptr_t
+findOriginalCxaThrowFunction(uintptr_t image_dli_fbase_address)
+{
+    SENTRY_ASYNC_SAFE_LOG_TRACE(
+        "Finding original __cxa_throw for image with base address %p", image_dli_fbase_address);
+
+    for (size_t i = 0; i < g_cxa_originals_count; i++) {
+        if (g_cxa_originals[i].image_dli_fbase_address == image_dli_fbase_address) {
+            return g_cxa_originals[i].cxa_throw_original_function;
+        }
+    }
+    SENTRY_ASYNC_SAFE_LOG_WARN("Address %p not found", image_dli_fbase_address);
+    return (uintptr_t)NULL;
+}
+
+static void
+addPair(SentryCrashImageToOriginalCxaThrowPair pair)
+{
+    uintptr_t originalCxaThrowFunction = findOriginalCxaThrowFunction(pair.image_dli_fbase_address);
+    if (originalCxaThrowFunction != (uintptr_t)NULL) {
+        SENTRY_ASYNC_SAFE_LOG_DEBUG("Already added address pair with image with base address: %p, "
+                                    "and originalCxaThrowFunction: %p",
+            (void *)pair.cxa_throw_original_function, (void *)pair.cxa_throw_original_function);
+        return;
+    }
+
+    SENTRY_ASYNC_SAFE_LOG_DEBUG(
+        "Adding pair for image with base address: %p, and originalCxaThrowFunction: %p",
+        (void *)pair.cxa_throw_original_function, (void *)pair.cxa_throw_original_function);
+
+    if (g_cxa_originals_count == g_cxa_originals_capacity) {
+        g_cxa_originals_capacity *= 2;
+        g_cxa_originals = (SentryCrashImageToOriginalCxaThrowPair *)realloc(g_cxa_originals,
+            sizeof(SentryCrashImageToOriginalCxaThrowPair) * g_cxa_originals_capacity);
+        if (g_cxa_originals == NULL) {
+            SENTRY_ASYNC_SAFE_LOG_ERROR(
+                "Failed to realloc memory for g_cxa_originals: %s", strerror(errno));
+            return;
+        }
+    }
+    memcpy(&g_cxa_originals[g_cxa_originals_count++], &pair,
+        sizeof(SentryCrashImageToOriginalCxaThrowPair));
+}
+
+static void
+__cxa_throw_decorator(void *thrown_exception, void *tinfo, void (*dest)(void *))
+{
+    const int k_requiredFrames = 2;
+
+    if (g_cxa_throw_handler != NULL) {
+        SENTRY_ASYNC_SAFE_LOG_DEBUG(
+            "Not __cxa_throw decorator, because no g_cxa_throw_handler set.");
+        g_cxa_throw_handler(thrown_exception, tinfo, dest);
```

Comment:
`m`: Is there a change that the `g_cxa_throw_handler` could call the original implementation inside, therefore calling it a second time at the end of this decorator method?